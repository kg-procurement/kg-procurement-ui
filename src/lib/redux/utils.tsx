import { SerializedError } from '@reduxjs/toolkit'
import { FetchBaseQueryError } from '@reduxjs/toolkit/query'

export type ZodParseError = Extract<
  FetchBaseQueryError,
  {
    status: 'PARSING_ERROR'
    originalStatus: number
    data: string
    error: string
  }
>

export function isZodParseError(
  err: SerializedError | FetchBaseQueryError | undefined,
): err is ZodParseError {
  if (!err) return false
  return (
    'data' in err &&
    'error' in err &&
    'originalStatus' in err &&
    'status' in err
  )
}

export function extractZodError(err: ZodParseError): React.ReactNode {
  const errorList = err.error.split('\n').slice(0, -1)
  return (
    <div className="flex flex-col">
      <div className="font-semibold">
        Response does not match validator schema:
      </div>
      <ul className="list-inside list-disc">
        {errorList.map((e, i) => (
          <li key={i}>{e}</li>
        ))}
      </ul>
    </div>
  )
}

export function isErrorWithMessage(
  err: FetchBaseQueryError,
): err is { status: number, data: { error: string } } {
  if (!err.data) return false
  return (
    typeof err.data === 'object' &&
    'error' in err.data &&
    typeof err.data.error === 'string'
  )
}
