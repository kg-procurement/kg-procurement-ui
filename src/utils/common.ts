export const noop = () => {}

export const formatPrice = (
  amount: number,
  quantityPrice: number,
  currencyCode: string,
  uomName: string
): string => {
  const formattedPrice = new Intl.NumberFormat(
    currencyCode === 'IDR' ? 'id-ID' : 'en-US',
    {
      style: 'currency',
      currency: currencyCode,
    }
  ).format(amount);

  return quantityPrice > 1
    ? `${formattedPrice} / ${quantityPrice} ${uomName}`
    : `${formattedPrice} / ${uomName}`;
};
