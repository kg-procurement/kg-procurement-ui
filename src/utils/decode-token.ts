export interface Claims {
  sub: string
  exp: number
}

export function decodeToken(token: string): Claims | null {
  try {
    const payload = JSON.parse(atob(token.split('.')[1]))
    return {
      sub: payload.sub,
      exp: payload.exp,
    }
  }
  catch (error) {
    console.error('Failed to decode token:', error)
    throw error
  }
}
