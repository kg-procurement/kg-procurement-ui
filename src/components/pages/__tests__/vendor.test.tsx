import { render, screen, waitFor } from '@testing-library/react'

import { withWrappers } from '@/lib/testing/utils.tsx'

import VendorPage from '../vendor.tsx'

describe('<VendorPage/>', () => {
  it('should render the header section with the logo', () => {
    render(withWrappers(<VendorPage />))

    const logo = screen.getByAltText('Kompas Gramedia Logo Background')
    expect(logo).toBeInTheDocument()
    expect(logo).toHaveAttribute('src', 'kompas-gramedia-logo-bg.svg')

    expect(screen.getByText('Search Vendor')).toBeInTheDocument()
    expect(screen.getByText('Lorem Ipsum')).toBeInTheDocument()
  })

  it('should render the vendors table content properly', async () => {
    const { container } = render(
      withWrappers(<VendorPage />, { withRoot: true }),
    )
    await waitFor(async () => {
      expect(screen.queryByTestId('loading-overlay')).not.toBeInTheDocument()
    })
    expect(container.innerText.trim()).toMatchSnapshot()
  })

  it('should render the footer', () => {
    render(withWrappers(<VendorPage />))
    expect(screen.getByText(/Â© 2024 KOMPAS/i)).toBeInTheDocument()
  })

  it('should render input boxes for filtering by product and location', () => {
    render(withWrappers(<VendorPage />))

    const productInput = screen.getByPlaceholderText('Filter by Product')
    const locationInput = screen.getByPlaceholderText('Filter by Location')

    expect(productInput).toBeInTheDocument()
    expect(locationInput).toBeInTheDocument()
  })
})
