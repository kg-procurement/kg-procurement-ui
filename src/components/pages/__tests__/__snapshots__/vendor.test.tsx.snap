// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<VendorPage/> > should render the vendors table content properly 1`] = `
"Evaluation Form
Search
Search Vendor
Lorem Ipsum

Vendor NameLocationPerformance ScoreAdriza Nikmah, PTIndonesia0...Toko AmazonIndonesia0...Aneka WarnaIndonesia0...Angkasa Buana Niaga Cipta, PTIndonesia0...Apartement Permata SenayanIndonesia0...Arta MuaraIndonesia0...Asaba Computer CentreIndonesia1...Astra Graphia TBK, PTIndonesia0...Asus Service CenterIndonesia0...Atrindo Asia Global, PTIndonesia0...
undefinedPrevious123undefinedMore pagesNext
undefined
0 of 99 row(s) selectedEmail Selected VendorÂ© 2024 KOMPAS
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
Contact us (Email):"
`;
